prize-tool/src/confirmation_latency.rs

```diff
@@ -0,0 +1,216 @@
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::process::exit;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+// Votes received `MAX_VOTE_DELAY` slots after the current slot will not be counted towards a
+// validator's latency score because this delay implies an availability issue rather than a latency
+// issue.
+const MAX_VOTE_DELAY: u64 = 10;
+
+#[derive(Default, Debug)]
+struct VoterEntry {
+    latency_score: i64, // +1 for low latency, -1 for high latency
+    last_slot: Slot,
+    last_hash: Hash,
+}
+
+// Checks `bank` voter state against the latest tracked `voter_record`. If voter hash has updated,
+// check if the voter has new votes to record. Returns a record of votes seen in this checkpoint
+// grouped by slot.
+fn voter_checkpoint(
+    bank: &Bank,
+    voter_record: &mut HashMap<Pubkey, VoterEntry>,
+) -> HashMap<Slot, HashSet<Pubkey>> {
+    let slot = bank.slot();
+    let mut slot_voters: HashMap<Slot, HashSet<Pubkey>> = HashMap::new();
+    for (voter_key, (_stake, account)) in bank.vote_accounts() {
+        let mut voter_entry = voter_record
+            .entry(voter_key)
+            .or_insert_with(VoterEntry::default);
+        if voter_entry.last_hash != account.hash {
+            voter_entry.last_hash = account.hash;
+            let vote_state = VoteState::from(&account).unwrap();
+            for lockout in vote_state.votes.iter().rev() {
+                if lockout.slot == voter_entry.last_slot {
+                    break;
+                }
+                if lockout.slot < slot.saturating_sub(MAX_VOTE_DELAY) {
+                    // vote was very late, don't track latency
+                } else {
+                    let voters = slot_voters.entry(lockout.slot).or_insert_with(HashSet::new);
+                    voters.insert(voter_key);
+                }
+            }
+            voter_entry.last_slot = vote_state.votes.back().unwrap().slot;
+        }
+    }
+    slot_voters
+}
+
+fn score_voters(voters: Vec<HashSet<Pubkey>>, voter_record: &mut HashMap<Pubkey, VoterEntry>) {
+    let total_voters: usize = voters.iter().map(|set| set.len()).sum();
+    let mut voters_seen = 0;
+    for voter_set in voters {
+        let is_low_latency = voters_seen < total_voters / 2;
+        let score_differential = if is_low_latency { 1 } else { -1 };
+        for voter in &voter_set {
+            if let Some(voter_entry) = voter_record.get_mut(&voter) {
+                voter_entry.latency_score += score_differential;
+            }
+        }
+        voters_seen += voter_set.len();
+    }
+}
+
+fn normalize_winners(winners: &[(Pubkey, f64)]) -> Vec<Winner> {
+    winners
+        .iter()
+        .map(|(key, latency_score)| (*key, format!("Latency score: {:.*}", 0, latency_score)))
+        .collect()
+}
+
+// Track voter latency by sequentially processing the blockchain specified by `block_slots` and
+// checkpointing the voter record after each entry.
+fn track_voter_latency(
```

Comment:
This fn looks like it could turn into a maintenance burden as it replicates a bunch of code from within `process_blocktree()`.  Could add a new option for a callback were `process_blocktree()` can call us at the right time to collect the vote info?